// moth/lib/personDB.js
var fs = require('fs')
	,	crypto = require('crypto')
	, client = require('redis').createClient()
	,	async = require('async')
	,	emailSpice = 'redtickcoonhound'
	,		path = require('path')
	,		resolvedPath = path.resolve(__dirname)
;

/*
 *	newPerson._id is email hashed
 */

client.select('people');

console.log(resolvedPath)

exports.verify = function(obj, cb){ // @obj === submited form data, req.body
	var hash = crypto.createHmac('sha1', emailSpice);
	
	var id = hash.update(obj.email).digest('base64')
		,	pw = obj.password
		,	cb = cb
	;
	
	async.parallel([checkRedis, checkHard], evaluate)
	
	function checkRedis(cb){
		client.hgetall(id, cb)
	};

	function checkHard(cb){
		fs.readFile('../peopleDB/' + id, 'utf8', cb)
	};

	function evaluate(err, result){
		if(err){cb(null, true); /*  no such user in hard db */ return}
		
		var hardData = JSON.parse(result[1]);
		
		if(result[0] && hardData){
			
				
			var user = 
					result[0].date_last_edited >= hardData.date_last_edited ?
					result[0] : 
					hardData 
				,	hash = crypto.createHmac('sha1', user.salt)
				,	hashedPass = hash.update(pw).digest('base64')
				;
				
			cb(null, (hashedPass === user.password) ? user : 'Wrong password')
			
		}
		
		else cb(null, false)
		
	};
	
};

exports.update = function(person){
	
};

exports.del = function(person){
	
};

exports.create = function(newPerson){
	
	try { // user email already exists // Also this ty catch shit is lame
	
		var p = fs.statSync('../peopleDB/' + newPerson._id); // if its not there, Node will throw up error
						
		return true
	
	}
	
	catch(err) { // no duplicate
				
		fs.writeFileSync('../peopleDB/' + newPerson._id, JSON.stringify(newPerson), 'utf8')
				
		client.hmset(newPerson._id, newPerson, function(e,r){if(e)console.log(e)})
		
		return false
		
	}
		
};
